raja@0.14.0 +rocm w/ ROCm 6.1.1: camp/resource/hip.hpp:114:21: error: no member named 'memoryType' in 'hipPointerAttribute_t'

### Steps to reproduce the issue

`raja@0.14.0 +rocm` build fails using:
* spack@develop (641075539c0a888d4378bd7dfdb19a872c5c42de from `Mon Jun 3 14:37:51 2024 -0700`)
* ROCm 6.1.1 installed via apt, externally defined via `packages:` config
* Ubuntu 22.04, X86_64

The reason `raja@0.14.0` is being built here is because we are trying to ultimately build `exago@1.6.0 +rocm` and the concretizer wants to use this older raja for it @cameronrutherford @pelesh @ryandanehy 

It looks like with ROCm >=6, the deprecated `memoryType` field of `hipPointerAttribute_t` was removed, see comments here:
* https://github.com/spack/spack/pull/44449#issuecomment-2146155133

Spack environment: [spack.yaml.txt](https://github.com/user-attachments/files/15685494/spack.yaml.txt)

<details><summary>Concretization</summary>

<pre>
 -   raja@0.14.0%gcc@11.4.0~cuda~desul+examples+exercises~ipo~omptask~openmp~plugins+rocm~run-all-tests+shared~tests~vectorization amdgpu_target=gfx90a build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-x86_64_v3
[+]      ^blt@0.4.1%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[+]      ^camp@0.2.3%gcc@11.4.0~cuda~ipo~openmp+rocm~tests amdgpu_target=gfx90a build_system=cmake build_type=Release generator=make patches=cb9e25b arch=linux-ubuntu22.04-x86_64_v3
[+]      ^cmake@3.20.6%gcc@11.4.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu22.04-x86_64_v3
[+]          ^curl@8.7.1%gcc@11.4.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu22.04-x86_64_v3
[+]              ^nghttp2@1.62.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^diffutils@3.10%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]              ^openssl@3.3.0%gcc@11.4.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^ca-certificates-mozilla@2023-05-30%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[+]              ^perl@5.38.2%gcc@11.4.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^berkeley-db@18.1.40%gcc@11.4.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^bzip2@1.0.8%gcc@11.4.0~debug~pic+shared build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[+]                  ^gdbm@1.23%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]                      ^readline@8.2%gcc@11.4.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu22.04-x86_64_v3
[+]              ^pkgconf@2.2.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]          ^ncurses@6.5%gcc@11.4.0~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=linux-ubuntu22.04-x86_64_v3
[+]          ^zlib-ng@2.1.6%gcc@11.4.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]      ^gcc-runtime@11.4.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[e]      ^glibc@2.35%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]      ^gmake@4.4.1%gcc@11.4.0~guile build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[e]      ^hip@6.1.1%gcc@11.4.0~cuda+rocm build_system=cmake build_type=Release generator=make patches=1f65dfe,6bbe83b arch=linux-ubuntu22.04-x86_64_v3
[e]      ^hsa-rocr-dev@6.1.1%gcc@11.4.0~asan+image+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-x86_64_v3
[e]      ^llvm-amdgpu@6.1.1%gcc@11.4.0~link_llvm_dylib~llvm_dylib+rocm-device-libs build_system=cmake build_type=Release generator=ninja patches=b4774ca arch=linux-ubuntu22.04-x86_64_v3
[e]      ^rocprim@6.1.1%gcc@11.4.0 amdgpu_target=auto build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-x86_64_v3
</pre></details>


Error:
```
$> spack -e . install /4c6p
...
==> Installing raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww [16/16]
==> No binary for raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww found: installing from source
==> No patches needed for raja
==> raja: Executing phase: 'initconfig'
==> raja: Executing phase: 'cmake'
==> raja: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

14 errors found in build log:
     199    /opt/rocm-6.1.1/bin/hipcc -c /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/src/PluginStrategy.cpp -o /t
            mp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_generated_PluginSt
            rategy.cpp.o -std=c++14 -D__HIP_PLATFORM_HCC__ -D__HIP_ROCclr__ -D__HIP_PLATFORM_AMD__ -D__HIP_PLATFORM_HCC__ -D__HIP_ROCclr__ -D__HIP_PLATFORM_AM
            D__ -fPIC -fPIC -I/tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/include -I/tmp/root/spack-stage/spack-s
            tage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/include -I/opt/rocm-6.1.1/include -I/opt/rocm-6.1.1/include -I/opt/rocm-6.1.
            1/include -I/opt/rocm-6.1.1/include -I/spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/camp-0.2.3-nvyy43ihj2fmyzz5muf6erpnq2dzcwca/include
     200    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/src/LockFreeIndexSetBuilders.cpp:24
            :
     201    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/include/RAJA/index/IndexSetBuilders
            .hpp:23:
     202    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/include/RAJA/index/IndexSet.hpp:23:
     203    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/include/RAJA/index/ListSegment.hpp:
            27:
     204    In file included from /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/camp-0.2.3-nvyy43ihj2fmyzz5muf6erpnq2dzcwca/include/camp/resource.hp
            p:28:
  >> 205    /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/camp-0.2.3-nvyy43ihj2fmyzz5muf6erpnq2dzcwca/include/camp/resource/hip.hpp:114:21: error: n
            o member named 'memoryType' in 'hipPointerAttribute_t'
     206      114 |           switch (a.memoryType) {
     207          |                   ~ ^
     208    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/src/AlignedRangeIndexSetBuilders.cp
            p:24:
     209    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/include/RAJA/index/IndexSetBuilders
            .hpp:23:
     210    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/include/RAJA/index/IndexSet.hpp:23:
     211    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/include/RAJA/index/ListSegment.hpp:
            27:
     212    In file included from /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/camp-0.2.3-nvyy43ihj2fmyzz5muf6erpnq2dzcwca/include/camp/resource.hp
            p:28:
  >> 213    /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/camp-0.2.3-nvyy43ihj2fmyzz5muf6erpnq2dzcwca/include/camp/resource/hip.hpp:114:21: error: n
            o member named 'memoryType' in 'hipPointerAttribute_t'
     214      114 |           switch (a.memoryType) {
     215          |                   ~ ^
     216    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/src/PluginStrategy.cpp:8:
     217    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/include/RAJA/util/PluginStrategy.hp
            p:11:
     218    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/include/RAJA/util/PluginContext.hpp
            :11:
     219    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/include/RAJA/policy/PolicyBase.hpp:
            21:
     220    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/include/RAJA/util/camp_aliases.hpp:
            32:
     221    In file included from /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/camp-0.2.3-nvyy43ihj2fmyzz5muf6erpnq2dzcwca/include/camp/resource.hp
            p:28:
  >> 222    /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/camp-0.2.3-nvyy43ihj2fmyzz5muf6erpnq2dzcwca/include/camp/resource/hip.hpp:114:21: error: n
            o member named 'memoryType' in 'hipPointerAttribute_t'
     223      114 |           switch (a.memoryType) {
     224          |                   ~ ^
     225    1 error generated when compiling for gfx90a.
     226    1 error generated when compiling for gfx90a.
     227    -- Removing /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_gene
            rated_LockFreeIndexSetBuilders.cpp.o
     228    /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/cmake-3.20.6-hlahh6mumfilfvjssuos7c6eddxygcx7/bin/cmake -E remove /tmp/root/spack-stage/sp
            ack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_generated_LockFreeIndexSetBuilders.cpp.o
     229    -- Removing /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_gene
            rated_AlignedRangeIndexSetBuilders.cpp.o
  >> 230    CMake Error at RAJA_generated_LockFreeIndexSetBuilders.cpp.o.cmake:199 (message):
     231      Error generating file
     232      /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_generated_Lock
            FreeIndexSetBuilders.cpp.o
     233
     234
  >> 235    make[2]: *** [CMakeFiles/RAJA.dir/build.make:94: CMakeFiles/RAJA.dir/src/RAJA_generated_LockFreeIndexSetBuilders.cpp.o] Error 1
     236    make[2]: *** Waiting for unfinished jobs....
     237    /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/cmake-3.20.6-hlahh6mumfilfvjssuos7c6eddxygcx7/bin/cmake -E remove /tmp/root/spack-stage/sp
            ack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_generated_AlignedRangeIndexSetBuilders.c
            pp.o
  >> 238    CMake Error at RAJA_generated_AlignedRangeIndexSetBuilders.cpp.o.cmake:199 (message):
     239      Error generating file
     240      /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_generated_Alig
            nedRangeIndexSetBuilders.cpp.o
     241
     242
  >> 243    make[2]: *** [CMakeFiles/RAJA.dir/build.make:80: CMakeFiles/RAJA.dir/src/RAJA_generated_AlignedRangeIndexSetBuilders.cpp.o] Error 1
     244    1 error generated when compiling for gfx90a.
     245    -- Removing /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_gene
            rated_PluginStrategy.cpp.o
     246    /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/cmake-3.20.6-hlahh6mumfilfvjssuos7c6eddxygcx7/bin/cmake -E remove /tmp/root/spack-stage/sp
            ack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_generated_PluginStrategy.cpp.o
     247    Generated /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_genera
            ted_MemUtils_CUDA.cpp.o successfully.
  >> 248    CMake Error at RAJA_generated_PluginStrategy.cpp.o.cmake:199 (message):
     249      Error generating file
     250      /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_generated_Plug
            inStrategy.cpp.o
     251
     252
  >> 253    make[2]: *** [CMakeFiles/RAJA.dir/build.make:122: CMakeFiles/RAJA.dir/src/RAJA_generated_PluginStrategy.cpp.o] Error 1
     254    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/src/MemUtils_HIP.cpp:23:
     255    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/include/RAJA/policy/hip/MemUtils_HI
            P.hpp:36:
     256    In file included from /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-src/include/RAJA/util/resource.hpp:23:
     257    In file included from /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/camp-0.2.3-nvyy43ihj2fmyzz5muf6erpnq2dzcwca/include/camp/resource.hp
            p:28:
  >> 258    /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/camp-0.2.3-nvyy43ihj2fmyzz5muf6erpnq2dzcwca/include/camp/resource/hip.hpp:114:21: error: n
            o member named 'memoryType' in 'hipPointerAttribute_t'
     259      114 |           switch (a.memoryType) {
     260          |                   ~ ^
     261    Generated /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_genera
            ted_MemUtils_SYCL.cpp.o successfully.
     262    1 error generated when compiling for gfx90a.
     263    -- Removing /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_gene
            rated_MemUtils_HIP.cpp.o
     264    /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/cmake-3.20.6-hlahh6mumfilfvjssuos7c6eddxygcx7/bin/cmake -E remove /tmp/root/spack-stage/sp
            ack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_generated_MemUtils_HIP.cpp.o
  >> 265    CMake Error at RAJA_generated_MemUtils_HIP.cpp.o.cmake:199 (message):
     266      Error generating file
     267      /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_generated_MemU
            tils_HIP.cpp.o
     268
     269
  >> 270    make[2]: *** [CMakeFiles/RAJA.dir/build.make:108: CMakeFiles/RAJA.dir/src/RAJA_generated_MemUtils_HIP.cpp.o] Error 1
     271    Generated /tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn/CMakeFiles/RAJA.dir/src/./RAJA_genera
            ted_DepGraphNode.cpp.o successfully.
     272    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn'
  >> 273    make[1]: *** [CMakeFiles/Makefile2:267: CMakeFiles/RAJA.dir/all] Error 2
     274    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-raja-0.14.0-4c6pdznhp3fzv53u72rqzc2k7q7p2tww/spack-build-4c6pdzn'
  >> 275    make: *** [Makefile:139: all] Error 2
```

### Error message

<details><summary>Error message</summary>

<pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.23.0.dev0 (136308c8fdc3d3d0593d563103f34e60ab15a5d6)
* **Python:** 3.10.12
* **Platform:** linux-ubuntu22.04-cascadelake
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/user-attachments/files/15685489/spack-build-env.txt)
[spack-build-out.txt](https://github.com/user-attachments/files/15685490/spack-build-out.txt)

@davidbeckingsale 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/camp/camp-rocm6.patch
+++ b/var/spack/repos/builtin/packages/camp/camp-rocm6.patch
@@ -1,0 +1,15 @@
+diff -ruN spack-src/include/camp/resource/hip.hpp spack-src-patched/include/camp/resource/hip.hpp
+--- spack-src/include/camp/resource/hip.hpp	2021-08-20 23:38:39.000000000 +0000
++++ spack-src-patched/include/camp/resource/hip.hpp	2024-06-06 21:34:25.174477941 +0000
+@@ -111,7 +111,11 @@
+         hipPointerAttribute_t a;
+         hipError_t status = hipPointerGetAttributes(&a, p);
+         if (status == hipSuccess) {
++#if (HIP_VERSION_MAJOR >= 6)
++          switch (a.type) {
++#else
+           switch (a.memoryType) {
++#endif
+             case hipMemoryTypeHost:
+               return MemoryAccess::Pinned;
+             case hipMemoryTypeDevice:

--- a/var/spack/repos/builtin/packages/camp/package.py
+++ b/var/spack/repos/builtin/packages/camp/package.py
@@ -54,6 +54,8 @@ class Camp(CMakePackage, CudaPackage, ROCmPackage):
 
     patch("libstdc++-13-missing-header.patch", when="@:2022.10")
 
+    patch("camp-rocm6.patch", when="@0.2.3 +rocm ^hip@6:")
+
     conflicts("^blt@:0.3.6", when="+rocm")
 
     def cmake_args(self):

